filter_test.go

```diff
@@ -143,6 +143,30 @@ file bar/foo
 `), b.String())
 }
 
+func TestWalkerIncludeReturnSkipDir(t *testing.T) {
+	d, err := tmpDir(changeStream([]string{
+		"ADD foo dir",
+		"ADD foo/x dir",
+		"ADD foo/y dir",
+		"ADD foo/x/a.txt file",
+		"ADD foo/y/b.txt file",
+	}))
+	assert.NoError(t, err)
+	defer os.RemoveAll(d)
+
+	found := []string{}
+
+	err = Walk(context.Background(), d, &FilterOpt{
+		IncludePatterns: []string{"**/*.txt"},
+	}, func(path string, info gofs.FileInfo, err error) error {
+		found = append(found, path)
+		return filepath.SkipDir
+	})
+	assert.NoError(t, err)
+
+	assert.Equal(t, []string{"foo"}, found)
```

Comment:
Why would this return anything at all. "foo" does not match the `IncludePatterns`.

Reply:
`Walk` finds a file that matches the include patterns, then starts calling `Walk` on parent directories, and the first `Walk` callback func returns `SkipDir`...

Reply:
Exactly that, yup.

Reply:
I don't understand how this makes sense. `IncludePatterns` is an option to `Walk`. Callback should not be called on paths that don't match `IncludePatterns`.

Reply:
If the callback calls are considered some "internal implementation specific behaviour" (what is should not I think) then I don't see point in giving the impression that the caller can control the behavior with `SkipDir` (as ultimately some internal logic is not controllable by parameters when determining if callbacks happen).

Reply:
If it's not included there then it isn't hooked up to Send/Recv - which means that we don't send the file stats for parent directories.

e.g. `foo` as this parent should have it's permissions/etc transferred through these methods. Potentially Send shouldn't use Walk at all, but that seems out of scope for this IMO, as it's a larger refactor and architectural change (not really convinced by it either).